runtime/src/main/java/org/corfudb/runtime/view/replication/IReplicationProtocol.java

```diff
@@ -63,67 +68,26 @@
-        return globalAddresses.parallelStream()
-                .map(a -> new AbstractMap.SimpleImmutableEntry<>(a, read(runtimeLayout, a)))
-                .collect(Collectors.toMap(r -> r.getKey(), r -> r.getValue()));
-    }
+    @Nonnull
+    Map<Long, ILogData> readAll(RuntimeLayout runtimeLayout, List<Long> addresses, boolean waitForWrite);
 
-    /** Read data from a range.
-     *
-     * <p>This method functions exactly like a readAll, except
-     * that it returns the result for a range of addresses.
-     *
-     * <p>An implementation may optimize for this type of
-     * bulk request, but the default implementation
-     * just performs multiple reads (possible in parallel).
-     *
-     * @param runtimeLayout         The RuntimeLayout stamped with layout to use for the readRange.
-     * @param globalAddresses       A list of addresses to read from.
-     * @return                      A map of addresses to committed
-     *                              addresses, hole filling if necessary.
-     */
-    Map<Long, ILogData> readRange(RuntimeLayout runtimeLayout, Set<Long> globalAddresses);
-    /** Peek data from a given address.
+    /**
+     * Peek data from a given address.
      *
      * <p>This function -may- return null if there was no entry
      * committed at the given global address, otherwise it
      * returns committed data at the given global address. It
      * does not attempt to hole fill if there was no entry.
      *
-     * @param runtimeLayout        The RuntimeLayout stamped with layout to use for the peek.
-     * @param globalAddress        The global address to peek from.
-     * @return                     The data that was committed at the
-     *                             given global address, or NULL, if
-     *                             there was no entry committed.
+     * @param runtimeLayout the RuntimeLayout stamped with layout to use for the peek.
+     * @param globalAddress the global address to peek from.
+     * @return the data that was committed at the given global address, or NULL, if
+     * there was no entry committed.
      */
     ILogData peek(RuntimeLayout runtimeLayout, long globalAddress);
-
-    /** Peek data from all the given addresses.
-     *
-     * <p>This method functions exactly like a peek, except
-     * that it returns the result for multiple addresses.
-     *
-     * <p>An implementation may optimize for this type of
-     * bulk request, but the default implementation
-     * just performs multiple peeks (possibly in parallel).
-     *
-     * @param runtimeLayout         The RuntimeLayout stamped with layout to use for the peekAll.
-     * @param globalAddresses       A set of addresses to read from.
-     * @return                      A map of addresses to uncommitted
-     *                              addresses, without hole filling.
-     */
-    default @Nonnull Map<Long, ILogData> peekAll(RuntimeLayout runtimeLayout,
```

Comment:
Hmm, after all it seems like this is not necessary, because we can just hole fill the prefix based on only peeking the greatest address to be read.